[inductor] parallel-compile: call triton_key() before forking

masnesral opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #127639

Summary:
A user reported severe slowdown on a workload when using parallel compile. The issue is that in some environments, the process affinity changes after forking such that all forked subprocesses use a single logical processor. Described here: https://github.com/pytorch/pytorch/issues/99625. That requires a separate fix, but during debuging we noticed that we can at least optimize the expensive call to triton_key() before forking.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

masnesral committed:
[inductor] parallel-compile: call triton_key() before forking

Summary:
A user reported severe slowdown on a workload when using parallel compile. I can't explain the underlying cause, but: There's a hash operation inside triton_key() that's very slow when executed in a forked process. The more forked processes, the slower. But strangely, I only see the slowdown if I import torch before forking. Importing within the forked processes instead is fast. Furthermore, it only happens in the user-provided environment (probably python 3.11).

And it gets more confusing... I have a simple repro; if I start two simultaneous instances of the test script, they run ~twice as slow. On the surface, it appears as though there's a mutex that's serializing all executions of hash operation.

While we investigate further, we can at least "fix" this specific slowdown by calling triton_key() before forking.

[ghstack-poisoned]

masnesral commented:
Note that we observe the same slowdown for both "fork" and "subprocess" (because the latter still uses a forked process pool inside the subprocess). This fixes the slowdown observed on both (at least on the user workload I was debugging)

eellison commented on torch/_inductor/async_compile.py:
```diff
@@ -45,6 +45,27 @@
+    try:
+        from triton.compiler.compiler import triton_key
+
+        triton_key()
```
i think triton key reads entire directory, so that could be why it's slow

anijain2305 reviewed (approved):
Thanks so much for debugging this.

masnesral commented:
@eellison, actually it hashes a .so, which seems to be kind of expensive, (e.g., 1s on my devgpu) but the slowdown we're seeing is like 10s of seconds, pretty much as if all sub-processes execute sequentially. But it only occurs on a very specific environment. Both @anijain2305 and I were unable to repro until I cloned the user's exact environment.

Here's a short script the repros:

```
import concurrent.futures
import hashlib
import time

# Importing torch before the fork causes slowdown:
import torch

def task(n):
    # No slowdown if we import torch here instead:
    # import torch

    path = "/home/slarsen/.conda/envs/pytorch-3.11/lib/python3.11/site-packages/triton/_C/libtriton.so"

    libtriton_hash = hashlib.sha256()
    with open(path, "rb") as f:
        while True:
            chunk = f.read(1024**2)
            if not chunk:
                break
            libtriton_hash.update(chunk)

    return True


def main():
    start = time.time()

    num_workers = 32
    with concurrent.futures.ProcessPoolExecutor(max_workers=num_workers) as executor:
        futures = {executor.submit(task, n): n for n in range(num_workers)}
        results = {}
        for future in concurrent.futures.as_completed(futures.keys()):
            results[futures[future]] = future.result()
        assert(len(results) == num_workers)
        #print(results)

    print(f"{time.time() - start:.2f}")


if __name__ == "__main__":
    main()
```

eellison commented:
@masnesral yea, the subprocess/fork stuff can be pretty sensitive to env. i think to repro the codecache import slowdown i needed to use same exact setup as our benchmark runners

masnesral commented:
This is how far I've gotten: the slowdown has something to do with CPU affinity. In the "bad" environment, the forked processes have a hard-coded CPU affinity of exactly two values: [0, 96]. In the "good" cases, the sub-processes are allowed to use the full set of CPUs. I've confirmed that the extra overhead observed in the "bad" cases is entirely in the (cpu-bound) hashing. I.e., it's not IO reading the triton .so. If I forcibly change the CPU affinity in the subprocesses, then the slowdown disappears. 

I can't figure out where the affinity change occurs. I've added logging to the Python multiprocessing library and the affinity in the parent process immediately before `os.fork()` is good (all cpus). The affinity in the child process immediately after fork, however, is bad. One possibility is env vars, but I also forcibly cleared the environment right before `os.fork()` and it makes no difference to the cpu affinity in the child. wut?

masnesral commented:
Ok, changing affinity after fork is a known issue: https://github.com/pytorch/pytorch/issues/99625
Verified that one of the workaround discussed there solves the problem (setting `KMP_AFFINITY=disabled`)

masnesral committed:
Update on "[inductor] parallel-compile: call triton_key() before forking"


Summary:
A user reported severe slowdown on a workload when using parallel compile. The issue is that in some environments, the process affinity changes after forking such that all forked subprocesses use a single logical processor. Described here: https://github.com/pytorch/pytorch/issues/99625. That requires a separate fix, but during debuging we noticed that we can at least optimize the expensive call to triton_key() before debugging.

[ghstack-poisoned]

masnesral committed:
Update on "[inductor] parallel-compile: call triton_key() before forking"


Summary:
A user reported severe slowdown on a workload when using parallel compile. The issue is that in some environments, the process affinity changes after forking such that all forked subprocesses use a single logical processor. Described here: https://github.com/pytorch/pytorch/issues/99625. That requires a separate fix, but during debuging we noticed that we can at least optimize the expensive call to triton_key() before forking.

[ghstack-poisoned]

masnesral commented:
@pytorchbot merge

masnesral commented:
@pytorchbot rebase

masnesral committed:
Update

[ghstack-poisoned]

masnesral commented:
@pytorchbot merge
